Create standard for mandatory and supported IaaS services (#587)
SovereignCloudStack/standards: Standards/scs-XXXX-vN-mandatory-and-supported-IaaS-services.md

---
title: Mandatory and Supported IaaS Services
type: Standard
status: Draft
track: IaaS
---

## Introduction

To be SCS-compliant a Cloud Service Provider (CSP) has to fulfill all SCS standards.

> Some of those standards are broad and consider all APIs of all services on the IaaS-Layer.

Reviewer: Can you please be specific and name at least one example?

There exist many services on that layer and for a first step they need to be limited to have a clear scope for the standards and the Cloud Service Providers following them.
For this purpose, this standard will establish lists for mandatory services that APIs have to be present in a SCS cloud as well as supported services, which APIS are considered by some standards and may even be tested for their integration but are optional in a sense that their omission will not violate SCS conformance.

## Motivation

There are many OpenStack APIs and their corresponding services that can be deployed on the IaaS level.
These services have differences in the quality of their implementation and liveness and some of them may be easily omitted when creating an IaaS deployment.
To fulfill all SCS-provided standards only a subset of these APIs are required.
Some more but not all remaining OpenStack APIs are also supported additionally by the SCS project and may be part of its reference implementation.
This results in different levels of support for specific services.

> This document will give readers insight about how the SCS classifies the OpenStack services accordingly.

Reviewer 2: not sure, but I guess we should also write APIs here?

```suggestion
This document will give readers insight about how the SCS classifies the OpenStack APIs accordingly.
```

If a cloud provides all mandatory and any number of supported OpenStack APIs, it can be tested for SCS-compliance.

> Any unsupported services will not be tested.

Reviewer 2: ```suggestion
Any unsupported APIs will not be tested.
```

## Mandatory IaaS APIs

The following IaaS APIs MUST be present in SCS-compliant IaaS deployments and could be implemented with the corresponding OpenStack services:

| Mandatory API | corresponding OpenStack Service | description |
|-----|-----|-----|
| **block-storage** | Cinder | Block Storage service |
| **compute** | Nova | Compute service |
| **identity** | Keystone | Identity service |
| **image** | Glance | Image service |
| **load-balancer** | Octavia | Load-balancer service |
| **network** | Neutron | Networking service |

> | **s3** or **object-store** | S3 API object storage | No formal standard exists, many implementations: Swift, RadosGW, minio... |

Reviewer 3: I personally dislike "or" here. There are user loads supporting only one of both so having here an OR does not bring much benefit. S3 support in it's own is not standardized enough to have multiple compatibility issues, especially when it comes to the OpenStack Identity related stuff.

Reviewer 3: We discussed this in the standardization SIG and following items were raised:
- S3 was always intended to be mandatory (also for KaaS layer)
- Swift should be at least "recommended"
Even though S3 and Swift both implement the same service (object-store) it can be seen as: HDMI output is a must and DisplayPort is recommended

Reviewer: And it should be two separate rows

Author: I will change the standard accordingly. 
**BUT**, if S3 is also listed as "object-store" equal to Swift, we have no chance to distinguish between those two services in the tests. 
This is a little bit unfortunate, because we will never know for certain, whether the endpoints are from an S3 service or from Swift. So I can change the lines here, but we will not be able to test them!

An option may be to use gaia-x credentials. Which raises the question: Do we have a point, were we state which credentials we require? @garloff ?

Reviewer 4: IMO, in this context, we need a standard for object storage too, similar like for block storage. CSPs need to know the requirements for object store and users need to know, which capabilities an object store has. See, #725

Reviewer 4: I agree to @josephineSei: We can not distinguish between two object store APIs in tests. We should define just one as mandatory.  Furthermore, setting S3 as mandatory is  problematic. As far as I known, S3 API is not in control of OpenStack. There is no official documentation on how S3 for a specific OpenStack release looks like. The latest documentation is from [Mitaka](https://docs.openstack.org/mitaka/config-reference/object-storage/configure-s3.html).  There is no documentation for (current release)[https://docs.openstack.org/2024.1/config-reference/object-storage/configure-s3.html], e.g. Hence, we standardize a moving target IMO.

Reviewer 4: There is still an open issue  with respect to version of a certain OpenStack service.  E.g. how does a consumer knows, which version of Nova is provided by a CSP? We will answer this question in #765

:::caution

S3 API implementations may differ in certain offered features.
CSPs must publicly describe, which implementation they use in their deployment.
Users should always research whether a needed feature is supported in the offered implementation.

:::

## Supported IaaS APIs

The following IaaS APIs MAY be present in SCS-compliant IaaS deployment, e.g. implemented thorugh the corresponding OpenStack services, and are considered in the SCS standards.

> | Mandatory API | corresponding OpenStack Service | description |

Reviewer 2: see discussion above

```suggestion
| API | corresponding OpenStack Service | description |
```

Author: I changed it to "Supported API"

|-----|-----|-----|
| **key-manager** | Barbican | Key Manager service |
| **billing** | Cloudkitty | Rating/Billing service |

> | **telemetry** | Ceilomete | Telemetry service |

Reviewer 5: ```suggestion
| **telemetry** | Ceilometer | Telemetry service |
```

| **dns** | Designate | DNS service |
| **time-series-databse** | Gnocchi | Time Series Database service |
| **orchestration** | Heat | Orchestration service |
| **bare-metal** | Ironic | Bare Metal provisioning service |
| **shared-file-systems** | Manila | Shared File Systems service |
| **ha** | Masakari | Instances High Availability service |

## Unsupported IaaS APIs

All other OpenStack services, which APIs are not mentioned in the mandatory or supported lists will not be tested for their compatibility and conformance in SCS clouds by the SCS community.
Those services MAY be integrated into IaaS deployments by a Cloud Service Provider on their own responsibility but the SCS will not assume they are present and potential issues that occur during deployment or usage have to be handled by the CSP on their own accord.

> The SCS standard offers no guarantees for compatibility or reliability of services categorized as unsupported in conjunction with an SCS-conformant infrastructure.

Reviewer 2: just personal opinion, I would shorten that sentence to make it easier to understand/read:

```suggestion
The SCS standard offers no guarantees for compatibility or reliability of services categorized as unsupported.
```

---
title: Mandatory and Supported IaaS Services
type: Standard
status: Draft
track: IaaS
---

## Introduction

To be SCS-compliant a CSP has to fulfill all SCS standards.
Some of those standards are broad and consider ALL services on the IaaS-Layer.
There exist many services on that layer and they need to be limited to have a clear scope for the standards and the Cloud Service Providers following them.
For this purpose, this standard will establish lists for mandatory services that have to be present in a SCS cloud as well as supported services, which are considered by some standards and may be tested or even implemented in the reference implementation but are optional in a sense that their omission will not violate SCS conformance.

## Motivation

There are many OpenStack APIs and their corresponding services that can be deployed on the IaaS level.
These services have differences in the quality of their implementation and liveness and some of them may be easily omitted when creating an IaaS deployment.
To fulfill all SCS-provided standards only a subset of these APIs are required.

> Some more but not all remaining OpenStack APIs are also supported additionally by the SCS project and may be part of its reference implementation.

Reviewer 4: I struggle a bit with

> ... and may be part of its reference implementation

I though recommended services **are** part of SCS reference implementation.

Author: I think we still need to discuss this. Some services were requested, but currently are not present in the reference implementation (Cloudkitty, Heat, Masakari, Skyline). Heat might still be excluded from the supported list.

Reviewer 6: Only Masakari is not available in the reference implementation at the moment. The other things are deployable, even if we would like to do without some of them. Masakari is also deployable, but we have not yet built and tested the necessary images.

Reviewer 2: Can this be marked as resolved?

This document will give readers insight about how the SCS looks at all the OpenStack services.
If a cloud provides all mandatory and maybe some supported OpenStack APIs and implementation of their services it can be tested for SCS-compliance.
Any unsupported services will not be tested.

## Mandatory OpenStack services

The following OpenStack services MUST be present in SCS-compliant IaaS-Deployments:

| OpenStack Service | description |
|-----|-----|
| **Cinder** | Block Storage service |
| **Glance** | Image service |
| **Keystone** | Identity service |
| **Neutron** | Networking service |
| **Nova** | Compute service |
| **Octavia** | Load-balancer service |
| **Placement** | Hardware Describing Service for Nova |
| **S3 API object storage** | No formal standard exists, many implementations: Swift, RadosGW, minio... |

:::caution

S3 API implementations may differ in certain offered features.
CSPs must publicly describe, which implementation they use in their deployment.
Users should always research whether a needed feature is supported in the offered implementation.

:::

## Supported OpenStack services

> The following services MAY be present in SCS-compliant IaaS-Deployment and they are considered in the SCS standards.

Reviewer 5: ```suggestion
The following services MAY be present in SCS-compliant IaaS deployment and are considered in the SCS standards.
```

> CSPs must publicly describe, which implementation they use in their deployment.

Reviewer 2: Should we maybe describe in more detail how this should be documented?
Maybe we should just enforce a machine-readable format for that, given there are not that many implementations in practical usage (mainly swift and ceph-rgw)?

That would have the benefit that we can easily test the compliance of this part of the standard.

Reviewer 2: Apologies, I was indeed hinting at making the CSPs description machine readable, as I already thought that it might not be possible to do this via openstack API, I should have written that down more clearly.

Reviewer 2: I don't have an idea or proposal on how to actually implement this and as nobody else has picked this up I'm going to mark this as resolved. I still think it would be a good idea to make this easily testable.

> Most of these services (except Cloudkitty, Gnocchi and Masakari) have been integrated and tested by the SCS reference implementation:

Reviewer 5: ```suggestion
```

This statement seems like a moving target to me, i.e. who is going to update this standard/statement if the component selection implemented in the reference setup changes?

I don't think this sentence adds any meaningful information to the standard that's why I'd remove it.

| OpenStack Service | description |
|-----|-----|
| **Barbican** | Key Manager service |
| **Cloudkitty** | Rating/Billing service |
| **Ceilometer** | Telemetry service |
| **Designate** | DNS service |
| **Gnocchi** | Time Series Database service |
| **Heat** | Orchestration service |
| **Horizon** | Dashboard |
| **Ironic** | Bare Metal provisioning service |
| **Manila** | Shared File Systems service |
| **Masakari** | Instances High Availability service |
| **Skyline** | Dashboard |

## Unsupported OpenStack services

> All other OpenStack services that are not mentioned in the mandatory or supported lists are not tested for their integration and behavior by the SCS community.

Reviewer 5: ```suggestion
All other OpenStack services that are not mentioned in the mandatory or supported lists are not tested for their compatibility and behavior in SCS clouds by the SCS community.
```

> All other OpenStack services that are not mentioned in the mandatory or supported lists are not tested for their compatibility and behavior in SCS clouds by the SCS community.

Reviewer 2: This wording could be read to imply that all supported services are tested for their compatibility and behavior, which I don't think is currently the case. If this is the case and I'm just unaware of the tests being done, please point me to these tests.

Also I'm not sure what is referred to when talking about "tested for their [..] behavior"? Does it mean the behavior of the service is SCS compliant - whatever that means, given that I'm not aware of any service specific standards with regards to the list of "supported" services? Maybe this sentence could be clarified a bit?

The only somewhat service specific standards I can think off of the top of my head are the image naming standard which could be considered to be glance specific and the entropy standard which could be viewed as nova specific, if you squint. So if there are service specific standards which we test against we should maybe list those here or at least link to them?

Author: In another comment @berendt mentioned that, except for Masakari, all supported services are available in the reference implementation. As I read his comment, they will eventually also test Masakari.

In deed there are standards that will and have to be applied to all used services: e.g. the role standard that is currently under development. This is also what I meant with behavior: if services have different role names or have a different meaning of role names, their behavior is different.
But i should rephrase this sentece, for better understanding.

Reviewer 2: Please don't conflate availability of services with "tested" services.
Not all services that are available to use are necessarily also tested.

Also I put tested in quotation marks, because it's really not defined anywhere what that actually means, beside SCS standard tests.

Reviewer 2: marking as resolved as we mostly talk about APIs now.

> Those services may be integrated into IaaS deployments by a CSP on their own responsibility but the SCS will not assume they are present and potential issues that occur during deployment or usage have to be handled by the CSP on their own accord.

Reviewer 5: ```suggestion
Those services MAY be integrated into IaaS deployments by a CSP on their own responsibility but the SCS will not assume they are present and potential issues that occur during deployment or usage have to be handled by the CSP on their own accord.
```

> Those services MAY be integrated into IaaS deployments by a CSP on their own responsibility but the SCS will not assume they are present and potential issues that occur during deployment or usage have to be handled by the CSP on their own accord.

Reviewer 4: ```suggestion
Those services MAY be integrated into IaaS deployments by a Cloud Service Provider on their own responsibility but the SCS will not assume they are present and potential issues that occur during deployment or usage have to be handled by the CSP on their own accord.
```

The SCS standard offers no guarantees for compatibility or reliability of services categorized as unsupported in conjunction with an SCS-conformant infrastructure.

## Related Documents

[The OpenStack Services](https://www.openstack.org/software/)

## Conformance Tests

The presence of the mandatory OpenStack services (except the S3) will be tested in a test-script.

> As S3 is a moving target, it may be integrated into the test, but will not let the Conformance test fail in general.

Reviewer 5: ```suggestion
As the S3 interface is a moving target, it may be integrated into the test suite but the test result will not be taken into account to determine conformance.
```

:::

## Supported IaaS APIs

> The following IaaS APIs MAY be present in SCS-compliant IaaS deployment, e.g. implemented thorugh the corresponding OpenStack services, and are considered in the SCS standards.

Author: No these APIs are truely optional, but in contrast to the unsupported APIs, we know they are likely to be used and such consider them in standards and (except for Masakari, if I understood @berendt correctly) even tested for integration.

I can edit the sentece to better reflect the wide but optional usage of these APIs, which is the reason for us to also think of them in standards and maybe sometimes test their integration?

> The following IaaS APIs MAY be present in SCS-compliant IaaS deployment, e.g. implemented thorugh the corresponding OpenStack services, and are considered in the SCS standards.

Author: Is it because of the last part of the sentence? "... are considered in the SCS standards"? This is, because we need (for some standards) to consider MANY possible cloud configurations. The role standard for example with the new RBAC roles must decide between old and new roles, which has to be rolled out throughout all services in a deployment. If at some point we want to use the new RBAC system, all servies need to be able to do this, and while the mandatory service already are in a good state, some supported services are not. SO we do have to consider all only supported services in this standard, because many clouds use them, and we don't want to break them.
The mandatory Services are more of a minimal list of services, that are needed to have an scs-compliant cloud.

I hope this explains it a little bit better.

> The following IaaS APIs MAY be present in SCS-compliant IaaS deployment, e.g. implemented thorugh the corresponding OpenStack services, and are considered in the SCS standards.

Reviewer 2: No, sorry, I think I didn't make clear what I meant:

In line 50 we have:

```suggestion
The following IaaS APIs MAY be present [..]
```

but in line 52, immediately following this we have:

```
| **Mandatory** API | corresponding OpenStack Service | description |
```

which imho is a contradiction in the heading.

> The following IaaS APIs MAY be present in SCS-compliant IaaS deployment, e.g. implemented thorugh the corresponding OpenStack services, and are considered in the SCS standards.

Author: Uh I made a mistake when adding the first column to the table. It should be "Supported API". Thank you for pointing it out, I changed it.
